### ISSUE TITLE
need a workaround for SLURM resource allocation bug

### ISSUE THREAD

**Issue by hppritcha** [ofi-cray/libfabric-cray#587]
When using libfabric GNI provider with native SLURM and without the --exclusive option,
there are certain cases where it seems that the Cray SLURM plugn(s) doesn't correctly
compute the number of FMA descriptors should be assigned to the job.  This results in
the GNI provider disabling itself since it doesn't think there are sufficient network resources.
We probably will need to make use of the `GNI_CDM_MODE_FMA_SHARED` cdm mode
option under these circumstances.  That's the default for craypich, hence it not seeing being
observed very frequently.

**Comment by hppritcha** [comment]
Turns out this is very hard to workaround.  The GNI FMA sharing logic seems to get messed up on the node where the underallocation of FMA descriptors is observed.  This causes `GNI_SmsgSendWTag` to always error out with `GNI_RC_ERROR_RESOURCE`.   This seems to require specific number of ranks on a node when using non-exclusive, the magic case seeming to be 5 ranks on a node.  Then 4 FMAS are allocated by the slurm plugin.

By setting `export UGNI_DEBUG=3`  for a test case, in this case
    srun --label -N2 -n 16 ./osu_allgather

One gets as output:

```
09: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[11]
07: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[12]
10: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[13]
03: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[14]
04: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[15]
00: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[16]
05: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[17]
02: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[18]
06: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[19]
08: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[20]
01: [    62]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fa2000) and RO flags (0x7ffff7fa1000) (0x7ffff7fa1010) into userspace, controlling FMAD[21]
13: [    54]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fba000) and RO flags (0x7ffff7fb9000) (0x7ffff7fb9010) into userspace
12: [    54]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fba000) and RO flags (0x7ffff7fb9000) (0x7ffff7fb9010) into userspace
15: [    54]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fba000) and RO flags (0x7ffff7fb9000) (0x7ffff7fb9010) into userspace
11: [    54]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fba000) and RO flags (0x7ffff7fb9000) (0x7ffff7fb9010) into userspace
14: [    54]   FMA: gni_fma_shared_init: Mapped RW state (0x7ffff7fba000) and RO flags (0x7ffff7fb9000) (0x7ffff7fb9010) into userspace
```

All the ranks on the node where there were not enough FMAS for #ranks are unable to get an FMA when initializing the setup in kgni.

**Comment by hppritcha** [comment]
adding external dependency label.  Cray slurm folks examining the resource allocation issue.

**Comment by ztiffany** [comment]
Cray internal bug 835162 is tracking a slurm issue that could be causing up to 5 instances to placed on a node with 1 core's limit of GNI resources.

**Comment by ztiffany** [comment]
schedmd bug 2343 tracks this:
http://bugs.schedmd.com/show_bug.cgi?id=2343

**Comment by hppritcha** [comment]
@ztiffany looks like schedmd fixed this bug.  could you update when cray thinks they will have this fix available for customers?

**Comment by sungeunchoi** [comment]
Fixed in Slurm 15.08.7 16.05.0pre1
